### ISSUE TITLE
dial limiting is racy

### ISSUE THREAD

**Issue by marten-seemann** [libp2p/go-libp2p#1105]
In tests we're seeing "Application Error 0x0" quite frequently, for example in the `TestNewDialOld`:
https://github.com/libp2p/go-libp2p/blob/a26ef706573a16ec04d2f0f2c3dcaab70c27fe40/p2p/host/basic/basic_host_test.go#L452-L484
The error occurs on the `Write` call. This is the error we'd expect from QUIC when a connection is closed by the application via `transport.Conn.Close()`.

From the server's perspective, this is inherently racy. Consider the following packet flow, visualizing the 3-way handshakes that both QUIC and TCP perform when setting up a connection.
<img width="598" alt="image" src="https://user-images.githubusercontent.com/1478487/118721672-03d85a80-b7e0-11eb-842d-96adc671dd35.png">
From the client's perspective, the TCP handshake finishes first, so it decides to use the TCP connection, and cancel the QUIC handshake. As the handshakes are run in separate threats, there's no guarantee that the QUIC handshake is canceled before QUIC's TLS Finished message is sent out, so it might happen that the QUIC connection is closed after the handshake completed.
From the server's perspective, the QUIC handshake finishes first, and it starts using the connection, only to notice that the connection is closed a short time later (which then causes the `Write` on the stream to error).

The fact that in practice TCP needs another round-trip for the cryptographic handshake means that we have to fit another round-trip in here for TCP. For the situation described above to occur, the RTTs experienced by QUIC and TCP need to conspire to make this sequence of events possible. This is unlikely and probably the reason we don't see this problem a lot more frequently.

**Comment by marten-seemann** [comment]
There are 2 ways we can deal with this:
1. Ignore it: In practice, this will barely happen. 
    * The QUIC handshake completes 1 RTT before the TCP handshake. 
    * We have no reason to believe that the RTT should be significantly different for TCP and for UDP packets. 
     * Regarding tests: we'd have to make sure that the client always speaks first, otherwise we risk flaky tests.
2. Fix the race:
    * While it's not a problem between TCP and QUIC, it might become a problem when we add a transport that takes the same number of round-trips as either TCP or QUIC.
    * To solve this race, we need an explicit wire signal from the client "I'm committing to **this** connection.". Only after receiving this signal, the server would consider the connection as fully set up.

Fortunately for us, we do have such a wire signal: The client will open a stream. I'm assuming that the client doesn't just dial the connection to "pre-heat" it, and stops immediately after the handshake.
We could then delay adding the connection to the swarm until we've accepted the first stream on that connection.

@Stebalien, wdyt? Also pinging @mxinden for his Rust perspective on this.

**Comment by Stebalien** [comment]
>  We could then delay adding the connection to the swarm until we've accepted the first stream on that connection.

That is an interesting approach but it could cause other problems (e.g., more connections). But it may be our best option and shouldn't really degrade things too much in practice?

Note: I think the _worst_ problem here is IPv6 + IPv4. We often try dialing both. But really, we should just work harder to not create a bunch of connections we won't use:

1. Ranked dials (https://github.com/libp2p/go-libp2p-swarm/pull/252).
2. Avoid advertising some addresses (e.g., if we _know_ our IPv4 addresses work for inbound connections, maybe just don't advertise v6?)

**Comment by marten-seemann** [comment]
> But it may be our best option and shouldn't really degrade things too much in practice?

In theory, the frame that opens the first stream could be sent in the same packet as the message that completes the handshake, so there'd be no delay at all. In practice, as handshake and opening the stream are done in separate go routines, we'll probably end up sending two packets, but the delay should still be minimal (unless the second packet is lost).

> That is an interesting approach but it could cause other problems (e.g., more connections). 

Why would there be more connections? The client still closes the connections that lost the race, so I wouldn't expect any significant change in the number of connections.

**Comment by mxinden** [comment]
> Also pinging @mxinden for his Rust perspective on this.

Thanks for the ping.

Do I understand correctly that go-libp2p tries to only have a single connection to a given peer at any time?

rust-libp2p does not implement a [happy eyeball](https://en.wikipedia.org/wiki/Happy_Eyeballs)-like strategy, partially due to not being prioritized, partially due to the fact that there is really only one transport supported today (TCP). QUIC is still [in progress](https://github.com/libp2p/rust-libp2p/pull/1334) and most projects using rust-libp2p still operate in a mostly-IPv4 environment (sadly).

In addition to the above the core of rust-libp2p exposes the notion of multiple connections instead of abstracting multiple connections into one. Concrete connections are explicit to protocols in the sense that e.g. a protocol like Kademlia is aware of the number of connections available and can actively choose which connection to use. When two or more connections to a given peer are established, none of them are actively closed, instead protocols may choose to use either number of conenctions with idle connections being eventually closed. Note: This might cause two connections to the same peer to persist, thus using more resources than ideally required.

> From the server's perspective, the QUIC handshake finishes first, and it starts using the connection, only to notice that the connection is closed a short time later (which then causes the Write on the stream to error).

In a case where a go-libp2p client establishes two connections to a rust-libp2p server, same race condition could occur. Though I am not sure how problematic it would be, given that the data being lost is by the server who did not intent to communicate in the first place.

> To solve this race, we need an explicit wire signal from the client "I'm committing to this connection.". Only after receiving this signal, the server would consider the connection as fully set up.
>
> Fortunately for us, we do have such a wire signal: The client will open a stream. I'm assuming that the client doesn't just dial the connection to "pre-heat" it, and stops immediately after the handshake.
We could then delay adding the connection to the swarm until we've accepted the first stream on that connection.

This sounds reasonable to me though I am not sure whether this race condition is severe enough to warrant the additional complexity. Would allowing multiple connections to the same peer, with idle connections being closed eventually, be an option go-libp3p as well?

**Comment by Stebalien** [comment]
> In theory, the frame that opens the first stream could be sent in the same packet as the message that completes the handshake, so there'd be no delay at all. In practice, as handshake and opening the stream are done in separate go routines, we'll probably end up sending two packets, but the delay should still be minimal (unless the second packet is lost).

Well.... multistream. We complete the "handshake" (on everything but QUIC) by sending a multistream header to select the multiplexer. This means that the receiver (server) learns that the connection has completed _first_ and could, in theory, start using it immediately.

> Why would there be more connections? The client still closes the connections that lost the race, so I wouldn't expect any significant change in the number of connections.

I.e., if the remote peer wants to open a connection _back_ (e.g., both sides are trying to "reform" a broken connection), we're more likely to end up with two connections the longer the actual connection process takes.

> Do I understand correctly that go-libp2p tries to only have a single connection to a given peer at any time?

"Tries". But not very hard. It basically just dials everything in parallel and see what comes out. However, we _do_ cancel other outbound connections once one has succeeded.

Note: we _used_ to cancel outbound connections when we received an inbound connection. We had to stop doing that because both sides would see inbound connections, then cancel their outbound connections (killing all connections).

Also note: We've introduced a concept of "transient" connections (relays) where some protocols can opt-in to using a transient connection, but most won't.

> This sounds reasonable to me though I am not sure whether this race condition is severe enough to warrant the additional complexity. Would allowing multiple connections to the same peer, with idle connections being closed eventually, be an option go-libp3p as well?

The issue here isn't really allowing multiple connections (they're allowed), it's the fact that we _cancel_ dials when one succeeds so we don't sit around waiting for a bunch of dials we don't need.

**Comment by marten-seemann** [comment]
> Well.... multistream. We complete the "handshake" (on everything but QUIC) by sending a multistream header to select the multiplexer. This means that the receiver (server) learns that the connection has completed first and could, in theory, start using it immediately.

Stream muxer selection happens before that. The swarm only deals with connections that are fully set up (i.e. security and multiplexer).
The logic I was suggesting would happen right here: https://github.com/libp2p/go-libp2p-swarm/blob/eef02f818639129a7184dba4ac0dbba1471a4b40/swarm_listen.go#L94-L107

**Comment by Stebalien** [comment]
Basically, in multistream, the receiver learns that the negotiation has
"finished" before the initiator. This means that that Accept call will
return before the dialer's dial call.

At the moment, this means the receiver can _immediately_ try to identify
and use the connection.

Now (where time counts half RTTs):

1. Time 0: Receiver sees a multiplexer protocol it supports (on multistream) and accepts it by echoing it back.
2. Time 0: Accept on the receiver returns.
3. Time 0: Receiver opens an identify stream. It only supports one protocol so it doesn't wait for any negotiation.
4. Time 1: Initiator sees the multiplexer protocol confirmation (0.5RTT later).
7. Time 1: Initiator initiates the identify protocol.
5. Time 1: Initiator sees the new identify stream.
6. Time 1: Initiator responds with identify packet.
7. Time 2: Receiver sees the identify response and can start opening streams.
8. Time 2: Receiver sees the identify request and responds.
9. Time 3: Initiator sees the identify response and can start opening streams.

After (where time counts half RTTs):

1. Time 0: Receiver sees a multiplexer protocol it supports (on multistream) and accepts it by echoing it back.
4. Time 1: Initiator sees the multiplexer protocol confirmation (0.5RTT later).
7. Time 1: Initiator initiates the identify protocol.
2. Time 2: Accept on the receiver returns.
8. Time 2: Receiver sees the identify request and responds.
3. Time 2: Receiver opens an identify stream. It only supports one protocol so it doesn't wait for any negotiation.
9. Time 3: Initiator sees the identify response and can start opening streams.
5. Time 3: Initiator sees the new identify stream.
6. Time 3: Initiator responds with identify packet.
7. Time 4: Receiver sees the identify response and can start opening streams.

Basically, the receiver loses a full RTT.

Now, does this matter? Usually no. Usually, the _initiator_ cares more
about latency than the receiver, so this is probably fine.
